addons/delivery_shipper/models/delivery_carrier.py

```diff
@@ -0,0 +1,245 @@
+
+        return {
+            'total_weight': original_weight_uom._compute_quantity(total_weight, target_weight_uom),
+            'estimated_dimension': estimated_dimension * 100,
+        }
+
+    def shipper_send_shipping(self, pickings, is_return=False):
+        shipper = self._get_shipper()
+        res = []
+        for picking in pickings:
+            if not picking.shipper_order_id:
+                # TODO: for "Send to Shipper" button after the delivery is validated and in state 'done' but cancelled
+                # need to add a check to trigger the create order here. The user should be allowed to do this
+                # can override the send_to_shipper to call create order first
+                raise UserError(_("You can only trigger a pickup after an order to Shipper is created. Please make sure to create it first."))
+
+            order_id = picking.shipper_order_id
+            if order_id:
+                order_result = {
+                    'exact_price': picking.shipper_order_amount,
+                    'tracking_number': order_id
+                }
+            res.append(order_result)
+            picking.carrier_tracking_ref = order_id
+
+            pickup_response = shipper._request_pickup(order_id)
+            logger.info(pickup_response)
+            picking.message_post(
+                body=_('Pickup request for %s was sent to Shipper with Order ID %s', picking.name, order_id)
+            )
+            label_response = shipper._get_shipping_label(order_id)
+            label_url = label_response.get('data', {}).get('url')
+            message_post = Markup(
+                '{label_text} <a href="{url}" target="_blank">{link_text}</a><br>'
+                "Order ID: {order_id}<br><br>"
+                "It's important to note that this URL will only be accessible for a period of three days, "
+                "after which it will expire. Please download the label within this timeframe to avoid any inconvenience."
+            ).format(
+                label_text=_("Get the shipping label:"),
+                url=label_url,
+                link_text=_("Here"),
+                order_id=order_id
+            )
+            picking.message_post(body=message_post)
+            picking.message_post(body=_("""Please print and stick the label on the package prior to handing it over to the
+                                        courier driver as it's mandatory."""))
+        return res
+
+    def shipper_create_order(self, pickings):
+        shipper = self._get_shipper()
+        for picking in pickings:
+            order = picking.group_id.sale_id
+            origin_id = self._shipper_get_area_id(origin=self.shipper_origin_address)
+            destination_id = self._shipper_get_area_id(order=picking)
+
+            rate_id = self.shipper_3pl_partner_id or order.order_line.filtered(lambda line: line.is_delivery)[-1].name.split()[0].strip('[]')
```

Comment:
Isn't this quite complicated?
Why would you need to find the sale order and parse the order line name to get the id?